src/Parser/AfflictionWarlock/Modules/Features/CooldownTracker.js

```diff
@@ -11,44 +13,91 @@ class CooldownTracker extends CoreCooldownTracker {
         BUILT_IN_SUMMARY_TYPES.DAMAGE,
       ],
     },
-    //for the sake of completeness, typically unused
+  ];
+
+  //for the sake of completeness, typically none of these are used
+  static castCooldowns = [
     {
       spell: SPELLS.SUMMON_INFERNAL_UNTALENTED,
+      duration: 25,
       summary: [
         BUILT_IN_SUMMARY_TYPES.DAMAGE,
       ],
     },
     {
       spell: SPELLS.SUMMON_DOOMGUARD_UNTALENTED,
+      duration: 25,
       summary: [
         BUILT_IN_SUMMARY_TYPES.DAMAGE,
       ],
     },
     {
       spell: SPELLS.GRIMOIRE_IMP,
+      duration: 25,
       summary: [
         BUILT_IN_SUMMARY_TYPES.DAMAGE,
       ],
     },
     {
       spell: SPELLS.GRIMOIRE_VOIDWALKER,
+      duration: 25,
       summary: [
         BUILT_IN_SUMMARY_TYPES.DAMAGE,
       ],
     },
     {
       spell: SPELLS.GRIMOIRE_SUCCUBUS,
+      duration: 25,
       summary: [
         BUILT_IN_SUMMARY_TYPES.DAMAGE,
       ],
     },
     {
       spell: SPELLS.GRIMOIRE_FELHUNTER,
+      duration: 25,
       summary: [
         BUILT_IN_SUMMARY_TYPES.DAMAGE,
       ],
     },
   ];
+
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+    const cooldownSpell = this.constructor.castCooldowns.find(cooldownSpell => cooldownSpell.spell.id === spellId);
+    if (cooldownSpell) {
+      // adding the fixed cooldown, now we need to remove it from activeCooldowns too
+      const cooldown = this.addFixedCooldown(cooldownSpell, event.timestamp);
+      this.activeCooldowns.push(cooldown);
+      debug && console.log(`%cCooldown started: ${cooldownSpell.spell.name}`, 'color: green', cooldown);
+    }
+    // super.on_byPlayer_cast(event) would call trackEvent anyway
+    this.trackEvent(event);
+  }
+
+  addFixedCooldown(cooldownSpell, timestamp) {
+    const cooldown = {
+      ...cooldownSpell,
+      start: timestamp,
+      end: timestamp + cooldownSpell.duration * 1000,
+      events: [],
+    };
+    this.pastCooldowns.push(cooldown);
+    return cooldown;
+  }
+
+  // on_event() might be more accurate but it would be most likely called much more
+  trackEvent(event) {
+    const finishedCooldowns = this.activeCooldowns.filter(cooldown => cooldown.end && cooldown.end < event.timestamp).map((_, index) => index);
```

Comment:
This would be cleaner:

```jsx
this.activeCooldowns
  .filter(cooldown => cooldown.end && cooldown.end < event.timestamp)
  .forEach((cooldown, index) => {
    debug && console.log(`%cCooldown ended: ${cooldown.spell.name}`, 'color: red', cooldown);
    this.activeCooldowns.splice(index, 1);
  });
```

Also might want to do `event.timestamp > cooldown.end` instead; it might be a little bit clearer that you're looking for cooldowns that have ended.

Reply:
Initially I thought about this too, I was worried it wouldn't work correctly (I remember that in C# if you iterate through a list with foreach and remove an element you'd get an exception that the collection changed while a iterating and I thought this would be the same case)

Reply:
Even better:

`this.activeCooldowns = this.activeCooldowns.filter(cooldown => cooldown.end && cooldown.end < event.timestamp);`